### ISSUE TITLE
CtTypeParameterReference cannot be used as actual type argument

### ISSUE THREAD

**Issue by pvojtechovsky** [INRIA/spoon#1238]
Problem: CtTypeParameterReference cannot be used as actual type argument.

Why? Because CtTypeParameterReference uses `parent` to resolve it's declaration. 
But the `parent` is changed by `CtTypeReference#addActualTypeArgument(CtTypeReference)`, so after it get's assigned it lost's it's declarer.

See test case, which can be found in PR #1237
```java
	public void testTypeParameterReferenceAsActualTypeArgument() throws Exception {
		CtType<Tacos> aTacos = buildNoClasspath(Tacos.class).Type().get(Tacos.class);
		
		CtTypeReference<?> typeRef = aTacos.getReference();
		CtTypeParameter typeParam = aTacos.getFormalCtTypeParameters().get(0);

		CtTypeParameterReference typeParamRef = typeParam.getReference();
		//typeParamRef is able to resolve it's typeParam - OK
		assertSame(typeParam, typeParamRef.getDeclaration());
		//the reference to parent is used by legacy 
		//CtTypeParameterReferenceImpl#getRecursiveDeclaration 
		//to detect declaration - it is probably wrong approach
		assertSame(typeParam, typeParamRef.getParent());
		
		//this assignment changes parent of typeParamRef to typeRef
		typeRef.addActualTypeArgument(typeParamRef);
		//stored typeParamRef is same like the added one, no clone - OK
		assertSame(typeParamRef, typeRef.getActualTypeArguments().get(0));
		//typeParamRef has got new parent 
		assertSame(typeRef, typeParamRef.getParent());

		//and therefore it lost the link to declaring CtFormalTypeDeclarer, 
		//so this assertion FAILS
		assertSame(typeParam, typeParamRef.getDeclaration());
	}
```
Solution?
I think `CtTypeParameterReference` should use `CtTypeReferenceImpl#declaringType` to reference the `CtFormalTypeDeclarer`, which declares it.
The problem is that type of `CtTypeReferenceImpl#declaringType` is `CtTypeReference` and not `CtFormalTypeDeclarer`.
So that solution would need some changes ... like:
* rename CtTypeReferenceImpl#declaringType to `declarer` and change the type to `CtFormalTypeDeclarer`
* add new method `CtType(Parameter?)Reference#getDeclarer()` with return value `CtFormalTypeDeclarer`
* add new method `CtType(Parameter?)Reference#setDeclarer(CtFormalTypeDeclarer)`
* throw exception in old CtTypeReferenceImpl#getDeclaringType() if declarer is not a type, but method or constructor

WDYT?

**Comment by pvojtechovsky** [comment]
Or what about CtTypeParameterReferenceImpl extends CtReferenceImpl? No property of CtTypeReferenceImpl makes sense for CtTypeParameterReference so it makes no sense to extend it. Then missing `declarer` property would be added to CtTypeParameterReference directly.

WDYT?

**Comment by pvojtechovsky** [comment]
There is question: **Does it makes sense to use CtTypeParameterReference out of scope of it's parent CtFormalTypeDeclarer?**
A) no - in this case we can keep current spoon model
B) yes - in this case the CtTypeParameterReference must reffer CtFormalTypeDeclarer

I expect that B) is right then it is question what member variable type to use to reference to CtFormalTypeDeclarer?
A) CtFormalTypeDeclarer
B) CtFormalTypeDeclarerReference (it would be a new interface)
C) CtActualTypeContainer (with new method `CtFormalTypeDeclarer getTypeParameterDeclarer()`
D) ?

So do we need a reference (variable which extends from CtReference) or can we use directly CtFormalTypeDeclarer?

Why to use reference? 
* Because on all other places in spoon mode is used CtTypeReference in similar cases
* May be there is case when we want to reference to CtFormalTypeDeclarer, which does not exist in the model (no class path mode, or an early state of spoon model building?)

Why not to use reference?
* because we NEVER need actual type arguments (CtActualTypeContainer) in case when we refer to declarer of CtTypeParameter - because the CtTypeParameter cannot be recursively parametrized in java.
* because the instance of CtFormalTypeDeclarer always exist in the model, so we do not need to "refer" to something what is not loaded.

WDYT?

**Comment by monperrus** [comment]
> in this case the CtTypeParameterReference must refer CtFormalTypeDeclarer

Yes, this relation makes sense. Normally we already have it using aTypeParamRef.getDeclaration().getTypeParameterDeclarer()

Correct?

**Comment by pvojtechovsky** [comment]
yes. But aTypParamRef.getDeclaration() fails in this case.

**Comment by monperrus** [comment]
where is the bug? in addActualTypeArgument or in getDeclaration?

**Comment by pvojtechovsky** [comment]
in spoon model

**Comment by monperrus** [comment]
do you mean in the implementation class CtTypeParameterReferenceImpl?

**Comment by pvojtechovsky** [comment]
I mean that after CtTypeParameterReferenceImpl is assigned to it's (arbitrary) parent then CtTypeParameterReferenceImpl is missing the field, which would point to anything what might be used to found a CtTypeParameter or it's CtFormalTypeDeclarer.

**Comment by monperrus** [comment]
OK, so the bug is in `addActualTypeArgument`.

**Comment by pvojtechovsky** [comment]
I do not agree. It is not trivial problem, which can be solved by change in one method. It needs a change in spoon model of CtTypeParameterReferenceImpl and then change in some methods.

**Comment by monperrus** [comment]
Do you mean adding a field to `CtTypeParameterReferenceImpl`? `declaringType` is not enough?

**Comment by pvojtechovsky** [comment]
no, because we do not need to point to `type` but to `declarer`, which can be type, method or constructor. So the  `declaringType` is bad name.

I see 2 solutions:
A) to add field `declarer`. But then there is question what type should have this variable. Please read this [comment](https://github.com/INRIA/spoon/issues/1238#issuecomment-289263967) where I already analyze details.
B) to add field 'declaration' of type `CtTypeParameter`, which would directly point to declaration of this reference. It is not common approach in spoon model. But may be in this case it is not so bad idea...

WDYT?

**Comment by monperrus** [comment]
> to add field |declarer|.
> to add field 'declaration'
The problem with both solutions is that it breaks the lookup-based design of references.

There must be a way  to dynamically lookup the declaration of a type parameter reference.

**Comment by pvojtechovsky** [comment]
So what do you suggest? If I would  know the solution, there would be PR for that :-)

**Comment by monperrus** [comment]
I'm having a look!

**Comment by monperrus** [comment]
closed with 5e7ee76
